sphinxext/icon.py

```diff
@@ -1,121 +1,322 @@
 import re
-from docutils import nodes
-from docutils.parsers.rst import roles
+import fett
+from docutils import nodes, statemachine
+from docutils.parsers.rst import roles, Directive
 from sphinx.util.nodes import split_explicit_title
+from docutils.utils.error_reporting import ErrorString
 
 PAT_STRIP = re.compile(r'-(?:o|(?:alt))$')
+ALL_SUFFIXES = ('-add', '-edit', '-exclamation', '-remove', '-restart')
+HELP_TEMPLATE = fett.Template('''
+.. list-table::
 
-# A pre-approved list of icons. Approval required to add to this list.
+{{ for name in icons }}
+   * - ``{{ name }}``
+     - :{{ role }}:`{{ name }}`
+{{ end }}
+''')
+
+# Each of the below classes describes an icon font. Each must provide:
+# * A name for the docutils icon node
+# * A name for the docutils role
+# * A prefix for the CSS class attached to the output <span>
+# * A set  of icon names
+# Approval required to add to icon lists.
 # Make sure you update the documentation in github.com:mongodb/docs-meta
 # as well when updating.
-ICONS = set([
-    'caret-right',
-    'caret-square-left',
-    'caret-square-right',
-    'check-circle',
-    'check-square',
-    'copy',
-    'edit',
-    'exclamation-circle',
-    'eye',
-    'globe',
-    'lock',
-    'pencil-alt',
-    'save',
-    'star',
-    'sync-alt',
-    'trash',
-    'trash-alt',
-    'users'
-])
-
-# Discouraged, but not prohibited (yet)
-ICONS.add('arrow-right')
-ICONS.add('asterisk')
-ICONS.add('check')
-ICONS.add('ellipsis-h')
-ICONS.add('minus')
-ICONS.add('plus')
-
-ICONSB = set([
-    'windows',
-])
+
+
+class IconsFA5:
+    """FontAwesome 5 Solid icon font"""
+    NODE_NAME = 'IconFA5'
+    ROLE_NAME = 'icon'
+    CSS_PREFIX = 'fas fa-'
+    ICONS = set([
+        'caret-right',
+        'caret-square-left',
+        'caret-square-right',
+        'check-circle',
+        'check-square',
+        'copy',
+        'edit',
+        'exclamation-circle',
+        'eye',
+        'globe',
+        'lock',
+        'pencil-alt',
+        'save',
+        'star',
+        'sync-alt',
+        'trash',
+        'trash-alt',
+        'users'
+    ])
+
+    # Discouraged, but not prohibited (yet)
+    ICONS.add('arrow-right')
+    ICONS.add('asterisk')
+    ICONS.add('check')
+    ICONS.add('ellipsis-h')
+    ICONS.add('minus')
+    ICONS.add('plus')
+
+
+class IconsBrandsFA5:
+    """FontAwesome 5 Brand icon font"""
+    NODE_NAME = 'IconBrandFA5'
+    ROLE_NAME = 'iconb'
+    CSS_PREFIX = 'fab fa-'
+    ICONS = set([
+        'windows',
+    ])
+
+
+class IconsFA4:
```

Comment:
Unfortunately, this is now impossible without breaking backwards compatibility

Reply:
Icons are used lightly in 10gen/baas-docs but very heavily in 10gen/mms-docs and 10gen/cloud-docs. The other repos do not use them at all.

Reply:
I think as we have already seen that not supporting cloud icons explicitly caused huge problems, not qualifying their use would also lock us into never being able to upgrade our other icon sets. If we qualify the cloud icons, we also liberate them from the rest of our icon usage and allow us to change cloud or the other repository icons at will. I would almost say we should abstract this further and great a logical cloud icon set that maps to a configurable font awesome icon set.

Reply:
It's not generally possible for us to build that kind of abstraction without knowing what sort of changes the MMS team might make in the future.

The best thing we can do IMO is give writers all of the needed icon options under clear namespaces, and let them deal with changing their usage as needed.

Put another way: we can't "upgrade" icon sets per se because that would be a breaking change

Reply:
ok agree as we discussed. the cloud icons are partially namespaced. the ones that aren't are explicitly referenced, and content writers will need to manage their deprecation if and when that happens. we can add icons required by cloud as needed from the fa4 icon set.